### DISCUSSION TITLE
How to update response schema in swagger for all endpoints after modifying response at middleware level in a FastAPI Application?

### DISCUSSION THREAD

**Discussion by raghavsikaria** [fastapi/fastapi#8459] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from starlette.types import ASGIApp, Receive, Scope, Send, Message
from starlette.requests import Request
import json
from starlette.datastructures import MutableHeaders
from fastapi import FastAPI


class MetaDataAdderMiddleware:
    application_generic_urls = ['/openapi.json', '/docs', '/docs/oauth2-redirect', '/redoc']
    
    def __init__(
            self, 
            app: ASGIApp
    ) -> None:
        self.app = app

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        if scope["type"] == "http" and not any([scope["path"].startswith(endpoint) for endpoint in MetaDataAdderMiddleware.application_generic_urls]):
            responder = MetaDataAdderMiddlewareResponder(self.app, self.standard_meta_data, self.additional_custom_information)
            await responder(scope, receive, send)
            return
        await self.app(scope, receive, send)


class MetaDataAdderMiddlewareResponder:

    def __init__(
            self,
            app: ASGIApp,
    ) -> None:
        """
        """
        self.app = app
        self.initial_message: Message = {}

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        self.send = send
        await self.app(scope, receive, self.send_with_meta_response)

    async def send_with_meta_response(self, message: Message):

        message_type = message["type"]
        if message_type == "http.response.start":
            # Don't send the initial message until we've determined how to
            # modify the outgoing headers correctly.
            self.initial_message = message

        elif message_type == "http.response.body":
            response_body = json.loads(message["body"].decode())

            data = {}
            data["data"] = response_body
            data['metadata'] = {
                'field_1': 'value_1',
                'field_2': 'value_2'
            }

            data_to_be_sent_to_user = json.dumps(data, default=str).encode("utf-8")

            headers = MutableHeaders(raw=self.initial_message["headers"])
            headers["Content-Length"] = str(len(data_to_be_sent_to_user))
            message["body"] = data_to_be_sent_to_user

            await self.send(self.initial_message)
            await self.send(message)


app = FastAPI(
    title="MY DUMMY APP",
)

app.add_middleware(MetaDataAdderMiddleware)

@app.get("/")
async def root():
    return {"message": "Hello World"}
```

### Description

So, I have created a Custom Middleware for my big FastAPI Application, which alters responses from all of my endpoints this way:

Response model is different for all APIs. However, my MDW adds meta data to all of these responses, in an uniform manner. This is what the **final response object** looks like:

```json
{
    "data": <ANY RESPONSE MODEL THAT ALL THOSE ENDPOINTS ARE SENDING>,
    "meta_data": 
    {
        "meta_data_1": "meta_value_1",
        "meta_data_2": "meta_value_2",
        "meta_data_3": "meta_value_3",
    }
}
```

So essentially, all original responses, are wrapped inside a `data` field, a new field of `meta_data` is added with all meta_data. This `meta_data` model is uniform, it will always be of this type:
```json
    "meta_data": 
    {
        "meta_data_1": "meta_value_1",
        "meta_data_2": "meta_value_2",
        "meta_data_3": "meta_value_3",
    }
```
Now the problem is, when the swagger loads up, it shows the original response model in schema and not the final response model which has been prepared. How to alter swagger to reflect this correctly?
I have tried this:

```python
# This model is common to all endpoints!
# Since we are going to add this for all responses 
class MetaDataModel(BaseModel):
    meta_data_1: str
    meta_data_2: str
    meta_data_3: str

class FinalResponseForEndPoint1(BaseModel):
    data: OriginalResponseForEndpoint1
    meta_data: MetaDataModel

class FinalResponseForEndPoint2(BaseModel):
    data: OriginalResponseForEndpoint2
    meta_data: MetaDataModel

and so on ...
```

This approach does render the Swagger perfectly, but there are 2 major problems associated with it:

+ All my FastAPI endpoints break and give me an error when they are returning response. For example: my endpoint1 is still returning the original response but the endpoint1 expects it to send response adhering to `FinalResponseForEndPoint1` model
+ Doing this approach for all models for all my endpoints, does not seem like the right way

### Operating System

Linux

### FastAPI Version

0.75.0

### Python Version

3.6.8

**Comment by jonatasoli** [comment]
Wouldn't this be the case to customize the output directly in the documentation instead of using the automatically generated documentation?

**Comment by rafsaf** [comment, marked as answer]
Maybe you could just extend it yourself? I doubt FastAPI will be ever so smart to detect such things in middlewares.

Read: https://fastapi.tiangolo.com/advanced/extending-openapi/

At the end of the day, you can think of generated openapi spec as a `dict`, you can generate the default output (the wrong one, actually), and process it in a way you want. This should be at least doable, but the question is an effort. But worth a try.
